### PR TITLE
FIREFLY-1042: Implement derived columns in tables

### PR DESCRIPTION
**Pull request by loitly**
https://jira.ipac.caltech.edu/browse/FIREFLY-1042

Unit test added to [src/firefly/test/edu/caltech/ipac/table/EmbeddedDbUtilTest.java](https://github.com/Caltech-IPAC/firefly/commit/68ed39b57ce2a11e6d7d7ffd9b20da7abf10c337#diff-0b106b584a60c816503ddfc61870810e71345a621a3ee840048074d8fbd5aea0)

Test instance: https://firefly-1042-derived-columns.irsakudev.ipac.caltech.edu/irsaviewer

BTW, there are 2 'add column' icons added to the toolbar because I cannot decide which one is better.  Please provide your preference.

Description taken from ticket:

- add a 'create new column' button to table toolbar with option to hide
  - popup dialog when clicked
  - Name, Expression, and Data Type are required fields with optional Units, Description, and UCD fields.
  - 'search' icon next to the Expression field will bring up Expression Creator similar to the Table's Advanced Filter component.
  - new column is created from an expression
  - column is added to the end and scrolled into view
  - rendered with a different color (maroon)
- (DERIVED_FROM=expression) is recorded and saved in the description of the column
  - this information is persisted when saved(ipac and votable), therefore will be recognized as derived when loaded again

Edited 11/4:
Based on @lrebull feedback, 
- added precision field when data type is double.  use votable's precision syntax
  - Fn, where n is the number of digits after the decimal
  - En, where n is the precision in scientific notation
  - precision cannot be modified once column is added.  We decided to not allow modification back then.
- make ucd a suggestion box populated with UCDList from v1.4

When saving as IPAC table, an exception is made to save DERIVED_FROM info.  Units, description, ucd and precision is not, because they are not a part of IPAC Table standards.  It's not hard to save them as Firefly's internal headers, but I don't know how CCB feels about it.

Edited 11/9:
Based on CCB feedback,
- mark required fields
- keep precision but make it optional
- use the + icon, remove the other one
- restrict column names to [a-z0-9_]
- fixed nullString to 'null'
- make ucd suggestion field multi-values aware

Clarification on double special values: NULL, NaN, +Inf, and -Inf

- If any part of the expression is NULL[*], the result is NULL.  This is true for any datatypes, in particular string, long/BIGINT, and DOUBLE.
- For DOUBLE only, in the absence of NULLs, if any part of the expression is NAN, or there is a division by zero, the result is NAN.

NULL, NaN, +Inf, and -Inf are supported values in the database as well as our table model on the server side.
However, when converting to JSON,  those values are all converted to NULL.  This is the limitation of our JSON library, both client and server.

@gpdf when you get a chance, will you provide suggestions for samples section in the Expression Creator popup?

### CODE REVIEW

**Comment by ejoliet** [comment]
I can't see the 'add' button in nightly IRSAViewer... is it only me?
Thanks!

**Comment by loitly** [comment]
>

**Comment by loitly** [comment]
> I can't see the 'add' button in nightly IRSAViewer... is it only me?

Sorry.  Build script error.  I'll fix then let you know.

**Comment by loitly** [comment]
@ejoliet fixed.  Test URL updated.

**Review by ejoliet — APPROVED** [review]
Great! That looks pretty good to me. 
I tested it, also saving a table and loading it. 

Couple of comments:
- of course now that we can add, can we edit? or remove? This is for future discussion and needs
- i liked the idea of the description "DERIVED from..."!
- I prefer the icon with a "+" on it

Thanks!!

**Comment by robyww** [comment]
I like this one..
image
But I am sure the preferences will be split

File: src/firefly/js/ui/PopupUtil.jsx [hunk 1]
```diff
@@ -62,22 +62,36 @@ export function showPinMessage(text) {
...
-          <PopupPanel title={title} modal={modal}>
-
-                  {content}
-
-          </PopupPanel>
-      );
-    DialogRootContainer.defineDialog(POPUP_DIALOG_ID, dialogContent);
-
-    dispatchShowDialog(POPUP_DIALOG_ID);
+      showPopup({ID: POPUP_DIALOG_ID, content, title, modal});
```
Comment (robyww) [inline src/firefly/js/ui/PopupUtil.jsx:66]: good cleanup.

**Review by robyww — APPROVED** [review]
It all looks good. It seems to works as expects. Saving includes the columns. It is a very nice feature.

- UI footprint is pretty low so I don't see a lot of UI discussion beyond the dialog.
   - We need feedback from scientist on the dialog.
   - Option stuff should probably be marked as `(optional)` (units, description,UCD)


- I like the `DERIVED_FROM=` in the description

**Comment by lrebull** [comment]
OK, this is **really** cool!!!

FWIW, I prefer the icon with the plus sign over the other icon. 

I **really** like the red headers and the fact that the derived nature of the columns is preserved even when the table is saved. I didn't even know you could do that, and I love it.

There is some apparently odd behavior with significant figures and I am sure that the others are going to complain; I am not sure if it is fixable (on short timescales or otherwise). Here is what I did and why I think it's weird (and what I think is happening):
1. search on ic417 over 1 degree in allWISE.
2. add a column that is functionally w1mpro-w2mpro (i made the name of the column w1mw2 because I know ipac tables hate dashes and underscores) and another that is w3mpro-w4mpro (name it w3mw4). 
3. the initial columns have way too many sig figs in both of them (see attached initial.png). BUT the first row of the w3-w4 column is just 4.871.
4. save the table and upload it back again into the tool. now, because the first row of the w3mw4 column has only 4.871, the rest of that entire column has only three decimal places (see attached final.png). This is an artifact of how the tool is casting the numbers, by taking a guess based on the first row. (I know this because there is at least one Jira ticket in the system  about this same issue.) 
As I say, I don't know if this is fixable (easily or otherwise), but I am 100% confident others will comment on it. :)

We will need to provide help to users regarding what "UCD" is (if for no other reason than I am not 100% sure what it is and how it differs from the other information here). 

We will need to do something to help users create IPAC table-compliant column headers. I deliberately made headers that are not compliant, saved the file, uploaded it back again, and it displayed the table no problem. But it will make other IRSA tools gag, I am confident. (There are Jira tickets related to this too, I know for a fact.)

Does it matter (for tbl file compliance if nothing else) that the places where the calculation cannot be made are empty, rather than 'null'?

How do I delete or change columns after creating them?

I'm sure I'll have more comments later. 


initial
final

**Comment by ejoliet** [comment]
Luisa, on item (4) - did you try to save votable or ipac table.
I thought the votable would save the formatting?


--
Emmanuel Joliet | Lead Applications Developer | Caltech/IPAC
1200 E California, MC 100-22 | Pasadena CA 91125 | Phone (626) 395-1489

On Nov 3, 2022, at 09:57, Luisa Rebull ***@***.***> wrote:

﻿

OK, this is really cool!!!

FWIW, I prefer the icon with the plus sign over the other icon.

I really like the red headers and the fact that the derived nature of the columns is preserved even when the table is saved. I didn't even know you could do that, and I love it.

There is some apparently odd behavior with significant figures and I am sure that the others are going to complain; I am not sure if it is fixable (on short timescales or otherwise). Here is what I did and why I think it's weird (and what I think is happening):

  1.  search on ic417 over 1 degree in allWISE.
  2.  add a column that is functionally w1mpro-w2mpro (i made the name of the column w1mw2 because I know ipac tables hate dashes and underscores) and another that is w3mpro-w4mpro (name it w3mw4).
  3.  the initial columns have way too many sig figs in both of them (see attached initial.png). BUT the first row of the w3-w4 column is just 4.871.
  4.  save the table and upload it back again into the tool. now, because the first row of the w3mw4 column has only 4.871, the rest of that entire column has only three decimal places (see attached final.png). This is an artifact of how the tool is casting the numbers, by taking a guess based on the first row. (I know this because there is at least one Jira ticket in the system about this same issue.)
As I say, I don't know if this is fixable (easily or otherwise), but I am 100% confident others will comment on it. :)

We will need to provide help to users regarding what "UCD" is (if for no other reason than I am not 100% sure what it is and how it differs from the other information here).

We will need to do something to help users create IPAC table-compliant column headers. I deliberately made headers that are not compliant, saved the file, uploaded it back again, and it displayed the table no problem. But it will make other IRSA tools gag, I am confident. (There are Jira tickets related to this too, I know for a fact.)

Does it matter (for tbl file compliance if nothing else) that the places where the calculation cannot be made are empty, rather than 'null'?

How do I delete or change columns after creating them?

I'm sure I'll have more comments later.

[initial]<https://user-images.githubusercontent.com/23038600/199782335-ad34dceb-3184-47d4-a30f-972a3e66ec95.png>
[final]<https://user-images.githubusercontent.com/23038600/199782339-bfa4a4e7-1deb-4e73-8066-fb7debba3cac.png>

**Comment by lrebull** [comment]
@ejoliet IPAC table. Yes, sorry, I was only talking about IPAC tables, because I was testing old-school stuff. ;)

**Comment by lrebull** [comment]
Ironically(?) the issue of the guessing at precision in IPAC tables is one of the tickets that Trey just archived: https://jira.ipac.caltech.edu/browse/FIREFLY-986

**Comment by lrebull** [comment]
I tried all the new stuff and it seems to work really well for me!! Looks great.

**Comment by lrebull** [comment]
> But, I need help with the Sample Expressions section in the Expression Creator popup.  You will see it when you click on the ‘search’ icon next to the Expression field.  If I don’t get any samples, I’ll have to remove that whole section.  Thanks.

Let's use this to sneak some applied advanced documentation in there (please check that my syntax is actually correct! these seem to work, as far as I know, if the catalog you've loaded has the corresponding columns):
w3mpro-w4mpro
sqrt(power(w3sigmpro,2) + power(w4sigmpro,2))
(ra-82.0158188)*cos(radians(dec))
phot_g_mean_mag-(5*log10(1000/parallax) - 5)

**Comment by lrebull** [comment]
ooh, new bug. You can't plot derived columns.

**Pull request merged**